Add to Numpy from Matrix Function

isaaclegred opened the pull request:
## Proposed changes

This PR introduces a function that allows a Matrix to be converted to a Numpy Array, and to be used in Python via Boost.  It will allow much easier access and manipulation of Data stored in Dat files, and will make Python examples about h5 features easier to follow in the future.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--

-->

isaaclegred committed:
Add function for converting a Matrix to a Numpy array in Python

nilsdeppe commented on src/DataStructures/Python/CMakeLists.txt:
```diff
@@ -9,7 +9,8 @@ spectre_python_add_module(
   Matrix.cpp
-  )
+  ToNumpyArray.cpp
+    )
```
Only two spaces here

nilsdeppe commented on src/DataStructures/Python/Matrix.cpp:
```diff
@@ -1,30 +1,36 @@
+
+#include <array>
 #include <boost/python.hpp>
+#include <boost/python/numpy.hpp>
```
We can't rely on this header existing and I don't think we need it since you use the numpy API directly

nilsdeppe commented on src/DataStructures/Python/Matrix.cpp:
```diff
@@ -1,30 +1,36 @@
+#include "DataStructures/Python/ToNumpyArray.hpp"
 #include "PythonBindings/BoundChecks.hpp"
 #include "Utilities/MakeString.hpp"
-
```
Please keep this blank line

nilsdeppe commented on src/DataStructures/Python/ToNumpyArray.hpp:
```diff
@@ -0,0 +1,13 @@
+// See LICENSE.txt for details.
+
+#pragma once
+#include <Python.h>
```
Please add a blank line between the `pragma once` and the `include`

nilsdeppe commented on tests/Unit/DataStructures/Test_Matrix.py:
```diff
@@ -4,7 +4,7 @@
 import unittest
 import math
-
+import numpy
```
Could you add a blank line after the import please?

nilsvu replied:
two blank lines, actually. You can use autopep8 (it's in the Docker container) to automatically format the file automatically

nilsdeppe commented on tests/Unit/DataStructures/Test_Matrix.py:
```diff
@@ -40,5 +40,13 @@ def assignment_test():
 
 
+    def test_to_numpy_array(self):
+        M = Matrix(2,2)
```
Could you please add spaces after the commas here?

nilsdeppe commented on src/DataStructures/Python/ToNumpyArray.cpp:
```diff
@@ -0,0 +1,31 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+#include "DataStructures/Python/ToNumpyArray.hpp"
```
Could you please add a blank line between the comment and the `include`?

nilsdeppe commented on src/DataStructures/Python/ToNumpyArray.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <array>
+#include <cstddef>
+#include <utility>
+#include "DataStructures/Python/Numpy.hpp"
```
Could you please make the include spacing as:
```cpp
#include <utility>

#include "DataStructures/Python/Numpy.hpp"
#include "DataStructures/Matrix.hpp"
```

nilsdeppe commented on src/DataStructures/Python/ToNumpyArray.cpp:
```diff
@@ -0,0 +1,31 @@
+namespace py_bindings {
+
+PyObject* to_numpy_array(const Matrix& matrix) {
+  double* c_style_data = static_cast<double*>(
```
Could you please add a comment to the effect of:
```cpp
// We use `malloc` instead of `new` because we tell NumPy it needs to free the memory
// and NumPy uses `free`, not `delete`.

nilsdeppe commented on src/DataStructures/Python/ToNumpyArray.cpp:
```diff
@@ -0,0 +1,31 @@
+      PyArray_SimpleNewFromData(2, dims.data(), NPY_DOUBLE, c_style_data);
+  PyArrayObject* npy_array_obj = reinterpret_cast<PyArrayObject*>(npy_array);
+  PyArray_ENABLEFLAGS(npy_array_obj, NPY_ARRAY_OWNDATA);
+  PyObject* npy_new_array = reinterpret_cast<PyObject*>(npy_array_obj);
```
I don't think you need this last cast, I think you can do:
```cpp
  PyArray_ENABLEFLAGS(npy_array_obj, NPY_ARRAY_OWNDATA);
  return npy_arry;
```
The reason is that after the first `reinterpret_cast` the pointer still points to the original object, and so you are mutating the original object that is also pointed to by `npy_array` (this is why we generally avoid pointers whenever possible :) )

nilsdeppe commented on src/DataStructures/Python/ToNumpyArray.cpp:
```diff
@@ -0,0 +1,31 @@
+      {static_cast<long>(matrix.rows()), static_cast<long>(matrix.columns())}};
+  PyObject* npy_array =
+      PyArray_SimpleNewFromData(2, dims.data(), NPY_DOUBLE, c_style_data);
+  PyArrayObject* npy_array_obj = reinterpret_cast<PyArrayObject*>(npy_array);
```
Could you add a comment along the lines of:
```cpp
// The `reinterpret_cast` is intentional because we know the pointer actually points
// to an object of type `PyArrayObject` and we need to access it in that manner.
```
The reason is that `reinterpret_cast` is one of the very dark corners of C++ and so it's good practice to state that you are intentionally using that feature and why.

nilsdeppe commented on src/DataStructures/Python/Numpy.hpp:
```diff
@@ -0,0 +1,12 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+#pragma once
```
Could you please add a blank line before and after the `pragma once`

nilsdeppe commented on src/DataStructures/Python/Matrix.hpp:
```diff
@@ -0,0 +1,10 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+#pragma once
```
Can you please add a blank line before the `pragma once`?

nilsdeppe commented on src/DataStructures/Python/Matrix.hpp:
```diff
@@ -0,0 +1,10 @@
+// See LICENSE.txt for details.
+#pragma once
+
+#include "DataStructures/Matrix.hpp"
```
You don't need either of these `include` statements because in the header you are not using either class

nilsdeppe commented on src/DataStructures/Python/Matrix.hpp:
```diff
@@ -0,0 +1,10 @@
+#include "Numpy.hpp"
+
+namespace py_bindings {
+void bind_Matrix();
```
Can you please make `Matrix` lower case? I.e. `bind_matrix`

nilsdeppe commented on tests/Unit/DataStructures/Test_Matrix.py:
```diff
@@ -40,5 +40,13 @@ def assignment_test():
+        A = M.to_numpy_array()
+        self.assertEquals(A[0,0], 1.0)
+        self.assertEquals(A[0,1], 2.0)
+        self.assertEquals(numpy.sin(A)[0,0], numpy.math.sin(1))
```
Could you please add a couple blank lines before the `if __name__ == '__main__':` statement?

nilsvu replied:
add two :)

nilsdeppe commented on tests/Unit/DataStructures/Test_Matrix.py:
```diff
@@ -40,5 +40,13 @@ def assignment_test():
+        M[0,0] = 1.0
+        M[0,1] = 2.0
+        A = M.to_numpy_array()
+        self.assertEquals(A[0,0], 1.0)
```
I think you can summarize all three asserts as `type(A) is numpy.ndarray`, or something similar

nilsvu replied:
`isinstance(A, np.ndarray)`

nilsdeppe reviewed (commented):
Here's some feedback. All are minor stylistic comments. Good job!

nilsvu commented on src/DataStructures/Python/Matrix.cpp:
```diff
@@ -36,10 +42,15 @@ void bind_matrix() {
+             self(bp::extract<size_t>(x[0]), bp::extract<size_t>(x[1])) = val;
+           })
+      // Conversion of Matrix to Numpy Array
+      .def("to_numpy_array",
```
Better name this `to_numpy` for consistency with, for instance, [Pandas](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_numpy.html). Also name the cpp file and function accordingly.

nilsvu replied:
Please add a Python docstring for this function, e.g. `Convert the Matrix to a Numpy array. Always creates a copy.` This is useful information since conversions to Numpy arrays often try to avoid copying.

nilsvu commented on src/DataStructures/Python/ToNumpyArray.cpp:
```diff
@@ -0,0 +1,31 @@
+// Distributed under the MIT License.
```
Please edit the commit message to be clearer, e.g. `Add function for converting a Matrix to a Numpy array in Python` or similar

nilsvu commented on src/DataStructures/Python/Bindings.cpp:
```diff
@@ -2,6 +2,14 @@
 #include <boost/python.hpp>
+// These macros are required so that the NumPy API will work when used in
+// multiple cpp files.
+#define PY_ARRAY_UNIQUE_SYMBOL SPECTRE_DATASTRUCTURES_PYTHON_BINDINGS
```
Please link to https://docs.scipy.org/doc/numpy/reference/c-api.array.html#importing-the-api here

nilsvu replied:
fix formatting here like you did in the other file

nilsdeppe replied:
@nilsleiffischer do you mean `// https` (add a space between slashes and the url)?

nilsvu commented on src/DataStructures/Python/Bindings.cpp:
```diff
@@ -2,6 +2,14 @@
+// Disable compiler warnings. NumPy ensures API compatibility among different
+// 1.x versions, as features become deprecated in Numpy 1.x will still function
+// but cause a compiler warning
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
```
To make this clearer, I think write the comment more like `Code is clean against Numpy 1.7` and link to https://docs.scipy.org/doc/numpy-1.15.1/reference/c-api.deprecations.html

nilsvu commented on src/DataStructures/Python/Numpy.hpp:
```diff
@@ -0,0 +1,12 @@
+#pragma once
+// These macros are required so that the NumPy API will work when used in
+// multiple cpp files.
+#define NO_IMPORT_ARRAY
```
Also edit the comments here (see above)

nilsvu commented on src/DataStructures/Python/ToNumpyArray.hpp:
```diff
@@ -0,0 +1,13 @@
+/// \endcond
+
+namespace py_bindings {
+PyObject* to_numpy_array(const Matrix& matrix);
```
Add the Python docstring (see above) also as a comment to this function, in particular that it always copies.

nilsvu reviewed (changes requested):
Have you looked into ways to make `Matrix` itself compatible with Numpy? Ideally we would be able to hand over a `Matrix` directly to Numpy functions without having to copy into a new `ndarray`. If we we plan on using these bindings we should give it a try now, and if we find this requires too much effort we can discard or defer it. A few notes on that:
- I think all we need is an implementation of the [`__array_ufunc__`](https://docs.scipy.org/doc/numpy-1.13.0/neps/ufunc-overrides.html) member function for `Matrix`. It tells Numpy what to do when it encounters a non-`ndarray`. Then, add the [`NDArrayOperatorsMixin`](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.lib.mixins.NDArrayOperatorsMixin.html) to get all the operator implementations for free.
- For an example, see [this implementation for a custom `Array` type in PyCBC](https://github.com/gwastro/pycbc/blob/master/pycbc/types/array.py#L205). If you replaced their `.numpy()` function with your `.to_numpy()` function, this would already do what we want, I think. The problem is that the data would be copied on every function call, which would be a lot of overhead. Another useful example is [this implementation in `pandas`](https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/numpy_.py#L175). Both of these implementations use `ndarray` as their underlying data type, which makes things easy.
- Closest to what we want could be [this implementation for the Armadillo library](https://sourceforge.net/p/armanpy/code/ci/master/tree/), but I didn't look into its details. If I understand correctly, their underlying data type is in C++ but the data is laid out the same as in `ndarray`.

nilsdeppe commented:
@nilsleiffischer the more long-term thing we were thinking is actually removing Matrix and having all C++ functions that interact via `Matrix` actually use NumPy array instead when we wrap them. What do you think about that?

nilsvu commented:
@nilsdeppe I think that would mean the data had to be copied between the cpp and the ndarray at every interface. So if we can make it work with `__array_ufunc__` that would probably be better

nilsdeppe commented:
Yes, though we don't use `Matrix` very often for at least what I think we will do with python and it's python, that alone means it'll be slow. However, I'm fine trying what you're saying with this `array_ufunc` thing. I don't think that should be considered as part of this PR because we need this in order to move forward on visualizing data.

nilsvu commented:
@nilsdeppe sure, it's an enhancement, created #1348. I'm not actually sure why we necessarily need to convert to Numpy arrays for visualisations. I have a bunch of code connecting spectre to yt, so I'd be happy discuss with whoever is working on these things.

nilsvu commented on src/DataStructures/Python/Matrix.cpp:
```diff
@@ -36,10 +42,14 @@ void bind_matrix() {
+             self(bp::extract<size_t>(x[0]), bp::extract<size_t>(x[1])) = val;
+           })
+      // Convert Matrix to a Numpy Array.  Always creates a copy.
+      .def("to_numpy", +[](const Matrix& self) { return to_numpy(self); });
```
My comment on this was probably not clear. I meant please add a Python docstring to this function, which is a string you pass as another argument here. It will be displayed as documentation when using the function in Python. Just look up what Python docstrings are if you are unfamiliar.

nilsdeppe replied:
@isaaclegred I found [this](https://stackoverflow.com/questions/6114462/how-to-override-the-automatically-created-docstring-data-for-boostpython) article and apparently you will need to do:
```cpp
.def("to_numpy", +[](const Matrix& self) { return to_numpy(self); },
     "Convert Matrix to a Numpy Array. Always creates a copy.");
```

nilsvu commented on src/DataStructures/Python/ToNumpy.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace py_bindings {
+
+// Convert Matrix to a Numpy Array.  Always creates a copy.
```
Move this documentation into the header file and format for Doxygen (the malloc stuff can stay in the implementation I think)

nilsvu commented on tests/Unit/DataStructures/Test_Matrix.py:
```diff
@@ -21,24 +22,30 @@ def test_getitem(self):
 
+    def test_to_numpy(self):
+        M = Matrix(2, 2)
+        M[0, 0] = 1.0
```
Add an off-diagonal entry to make sure the data layout is correct

nilsdeppe replied:
probably safest to set all elements to values like 1.4, 4.3, 7.12, -9.35 (they aren't multiples of each other) and test that.

nilsdeppe commented on src/DataStructures/Python/ToNumpy.cpp:
```diff
@@ -0,0 +1,38 @@
+namespace py_bindings {
+
+// Convert Matrix to a Numpy Array.  Always creates a copy.
+// We use `malloc` instead of `new` to tell NumPy it needs to free the memory
```
not `to tell NumPy` but `because we tell`

kidder commented:
@nilsleiffischer please look at the fixups

nilsvu commented on src/DataStructures/Python/Numpy.hpp:
```diff
@@ -0,0 +1,14 @@
+// Distributed under the MIT License.
```
What’s the purpose of this separate file? If it is to bundle the `define`s, then also use it in `Bindings.cpp`

isaaclegred replied:
I believe that the `Bindings.cpp` file is unique because it initializes the Python Module,  so it cannot have the `#define NO_IMPORT_ARRAY`, as opposed to all of the other files which use the Array API, which need to have `NO_IMPORT_ARRAY` defined, see https://docs.scipy.org/doc/numpy-1.15.1/reference/c-api.array.html, although @nilsdeppe would be better equipped to answer this than me.

nilsdeppe replied:
The reason for originally preferring what is done now is exact what @isaaclegred said. I didn't really want:

```cpp
#define NO_IMPORT_ARRAY
#include "Numpy.hpp"
```

since this does the weird thing of controlling the code generated in `Numpy.hpp` with a local macro. This is an antipattern that will blow up when C++ modules lands and so I think should be avoided.

nilsvu replied:
makes sense 👍

nilsdeppe replied:
@isaaclegred could you please add a comment like:
```cpp
// We do not use the antipattern:
// #define NO_IMPORT_ARRY
// #include "DataStructures/Python/Numpy.hpp"
// because
// 1. This means we are controlling code in the header file with a
//    local macro, which can be very confusing and should generally be avoided.
// 2. When C++ modules land this type of pattern will no longer really be possible
//    since it goes against exactly what modules is trying to do: make things
//    modular.
```
to where the Bindings.cpp file?

We should probably also move `numpy.hpp` into `src/PythonBindings` so other people will find it :)

I'm fine with you squashing all this.

isaaclegred replied:
Sounds good, will do.

nilsvu reviewed (commented):
Squash please

nilsdeppe commented on src/DataStructures/Python/Bindings.cpp:
```diff
@@ -3,13 +3,34 @@
+// 1. This means we are controlling code in the header file with a
+//    local macro, which can be very confusing and should generally be avoided.
+// 2. When C++ modules land this type of pattern will no longer really be
+// possible
```
nitpick on formatting here

nilsdeppe reviewed (commented):
Looks good, one nitpick on formatting of a new comment. I'm happy for you to squash. I think this is ready once the formatting is adjusted.

@nilsleiffischer could you please look at this?

nilsvu commented on src/DataStructures/Python/Bindings.cpp:
```diff
@@ -3,13 +3,34 @@
 #include <boost/python.hpp>
 
+// These macros are required so that the NumPy API will work when used in
+// multiple cpp files.
```
Add `See` before the link (like in the other file)

nilsvu reviewed (changes requested):
LGTM 👍 While you're squashing, just fix this one nitpick, too.

nilsdeppe reviewed (dismissed):
@nilsleiffischer could you please look at this?

kidder commented:
@isaaclegred this is failing to compile

isaaclegred commented:
Yeah sorry, it seems I forgot to change the path in the include in `ToNumpy.cpp` when I moved `Numpy.hpp` back to `DataStructures/Python`, I've changed it now, and checked that it compiled.

kidder commented:
@nilsleiffischer this needs to be looked at again
